src/renderer/components/+extensions/extensions.tsx

```diff
@@ -32,27 +54,238 @@ export class Extensions extends React.Component {
+    const pendingFiles: Promise<any>[] = [];
+
+    // read extensions with provided system path if any
+    installRequests.forEach(ext => {
+      if (ext.data) return;
+      const promise = fse.readFile(ext.filePath)
+        .then(data => ext.data = data)
+        .catch(err => {
+          Notifications.error(`Error while reading "${ext.filePath}": ${String(err)}`);
+        });
+      pendingFiles.push(promise);
+    });
+    await Promise.all(pendingFiles);
+    installRequests = installRequests.filter(item => item.data); // remove items with reading errors
+
+    // prepare temp folder
+    const tempFolder = path.join(os.tmpdir(), "lens-extensions");
+    await fse.ensureDir(tempFolder);
+
+    // copy files to temp, get extension info from package.json and do basic validation
+    const validatedInstalls: Promise<InstallRequestValidated>[] = installRequests.map(async installReq => {
+      const { fileName, data } = installReq;
+      const tempFile = path.join(tempFolder, fileName);
+      await fse.writeFileSync(tempFile, data); // copy to temp
+      try {
+        const packageJson: Buffer = await readFileFromTar(tempFile, {
+          // tarball from npm contains single root folder "package/*"
+          fileMatcher: (path: string) => !!path.match(/(\w+\/)?package\.json$/),
+          notFoundMessage: "Extension's manifest file (package.json) not found",
+        });
+        const manifest: LensExtensionManifest = JSON.parse(packageJson.toString("utf8"));
+        if (!manifest.lens && !manifest.renderer) {
+          throw `package.json must specify "main" and/or "renderer" fields`;
+        }
+        return {
+          ...installReq,
+          manifest: manifest,
+          tmpFile: tempFile,
+        };
+      } catch (err) {
+        fse.unlink(tempFile).catch(() => null); // remove invalid temp file
+        Notifications.error(
+          <div className="flex column gaps">
+            <p>Installing <em>{fileName}</em> has failed, skipping.</p>
+            <p>Reason: <em>{String(err)}</em></p>
+          </div>
+        );
+      }
+    });
+
+    // final step, provide UI with extension info for reviewing and confirming installation
+    const extensions = await Promise.all(validatedInstalls);
+    extensions.forEach(install => {
+      if (!install) {
+        return; // skip validating errors if any
+      }
+      const { fileName, manifest } = install;
+      const { name, version, description } = manifest;
+      const extensionFolder = this.getExtensionDestFolder(name);
+      const folderExists = fse.existsSync(extensionFolder);
+      const removeNotification = Notifications.info(
+        <div className="InstallingExtensionNotification flex gaps align-center">
+          <div className="flex column gaps">
+            <p>Install extension <b title={fileName}>{name}@{version}</b>?</p>
+            <p>Description: <em>{description}</em></p>
+            {folderExists && (
```

Comment:
We should probably show this notification only if `folderExists`. Extension should be also automatically enabled after install.

Reply:
`foldersExists &&` does what you say, extension is enabled by default (if was not disabled previously)

Reply:
> `foldersExists &&` does what you say, extension is enabled by default (if was not disabled previously)

This notification still always prompts user to hit "Install" button, that's unnecessary unless folder exists.

Reply:
How otherwise you can re-install/update?

Reply:
> Oh, or do you mean install automatically if "all good and not installed yet"? And show notification when exists?

Yes 🙂